### PR TITLE
SAM: Sharpness-Aware Minimization for flat minima

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model converges to sharp minima that perform well in-distribution but poorly on OOD inputs (p_oodc gap of 1.04 vs ensemble, p_re gap of 0.50). **Sharpness-Aware Minimization (SAM)** (Foret et al., ICLR 2021) explicitly seeks flat minima by optimizing for the worst-case loss in a neighborhood of the current weights.

SAM's update rule:
1. Compute gradient: g = nabla L(w)
2. Compute adversarial perturbation: epsilon = rho * g / ||g||
3. Compute gradient at perturbed point: g' = nabla L(w + epsilon)
4. Update: w = w - lr * g'

This forces the optimizer to find parameters where the loss is low AND the loss surface is flat (robust to perturbation). Flat minima generalize better because small input distribution shifts don't dramatically change the output.

**Why SAM specifically for this problem:**
1. **OOD gap is our main weakness.** p_oodc (7.643) and p_re (6.300) have the biggest gaps vs the 16-seed ensemble. SAM's flat minima directly target OOD robustness.
2. **Warm restarts just failed (#2238)** trying to find better basins by re-exploring. SAM takes the opposite approach: don't escape the current basin, but find the FLATTEST point within it.
3. **Proven in vision transformers.** SAM improves ViT generalization by 1-3% on ImageNet. Our Transolver is architecturally similar (attention-based, similar depth).
4. **Compatible with Lion.** SAM wraps the existing optimizer — it uses Lion for the actual weight update, just with a perturbed gradient.

**Trade-off:** SAM requires 2 forward+backward passes per step (~2x compute). With ~150 epochs normally, expect ~75-80 epochs. Adjust cosine_T_max to 80 to complete the schedule.

**Confidence:** Medium-High. SAM is one of the most well-validated generalization techniques in modern ML. Main risk: 2x compute reduces epoch count, and the model may need the full 150 epochs to converge.

**Reference:** Foret et al., "Sharpness-Aware Minimization for Efficiently Improving Generalization," ICLR 2021. https://arxiv.org/abs/2010.01412

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flags

```python
sam: bool = False          # Sharpness-Aware Minimization
sam_rho: float = 0.05      # perturbation radius (neighborhood size)
```

### Step 2: Implement SAM step

The SAM step replaces the standard `optimizer.step()`. Implement it as a function:

```python
def sam_step(model, optimizer, loss_fn, batch, cfg):
    """One SAM optimization step: ascend then descend."""
    # Step 1: Compute gradient at current point
    loss = loss_fn(model, batch)
    loss.backward()
    
    # Step 2: Save current parameters and compute perturbation
    old_params = {}
    with torch.no_grad():
        grad_norm = 0.0
        for name, p in model.named_parameters():
            if p.grad is not None:
                grad_norm += p.grad.data.norm(2).item() ** 2
            old_params[name] = p.data.clone()
        grad_norm = grad_norm ** 0.5
        
        # Perturb: w + rho * grad / ||grad||
        for name, p in model.named_parameters():
            if p.grad is not None:
                eps = cfg.sam_rho * p.grad.data / (grad_norm + 1e-12)
                p.data.add_(eps)
    
    # Step 3: Compute gradient at perturbed point
    optimizer.zero_grad()
    loss_sam = loss_fn(model, batch)
    loss_sam.backward()
    
    # Step 4: Restore original parameters and apply the perturbed gradient
    with torch.no_grad():
        for name, p in model.named_parameters():
            p.data.copy_(old_params[name])
    
    # Step 5: Standard optimizer step with the perturbed gradient
    optimizer.step()
    
    return loss.item()
```

**IMPORTANT implementation notes:**
- The `loss_fn(model, batch)` function should encapsulate the FULL training loss (surface pressure + volume + velocity + PCGrad + DCT freq — everything).
- You need to refactor the existing training step into a callable `loss_fn` that takes the model and batch and returns the scalar loss. This may require extracting the loss computation into a separate function.
- If this refactoring is too complex, a simpler approach: just double the forward/backward pass inline in the training loop with the perturbation logic between them.

### Step 3: Integrate into training loop

Replace the standard training step with the SAM step when `cfg.sam` is enabled. The key change is that each training step now does TWO forward+backward passes instead of one.

**If refactoring loss_fn is too complex**, use this inline approach:
```python
if cfg.sam:
    # First forward+backward (compute gradient at current point)
    loss = <existing loss computation>
    loss.backward()
    
    # Compute perturbation
    with torch.no_grad():
        grad_norm = sum(p.grad.data.norm(2)**2 for p in model.parameters() if p.grad is not None) ** 0.5
        old_params = {n: p.data.clone() for n, p in model.named_parameters()}
        for p in model.parameters():
            if p.grad is not None:
                p.data.add_(cfg.sam_rho * p.grad.data / (grad_norm + 1e-12))
    
    # Second forward+backward (gradient at perturbed point)
    optimizer.zero_grad()
    loss_sam = <same loss computation again>
    loss_sam.backward()
    
    # Restore and step
    with torch.no_grad():
        for n, p in model.named_parameters():
            p.data.copy_(old_params[n])
    optimizer.step()
else:
    # Standard step (existing code)
    loss.backward()
    optimizer.step()
```

### Step 4: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/sam-s42" \
  --wandb_group "round19/sam-optimizer" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 80 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --sam --sam_rho 0.05

# Seed 73 — identical but --seed 73 --wandb_name "frieren/sam-s73"
```

Key changes: `--sam --sam_rho 0.05 --cosine_T_max 80`

### Step 5: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, baseline comparison, W&B run IDs.

Report:
- **Epoch count**: Should be ~70-80 (half of baseline due to 2x compute). If much less, SAM overhead is higher than expected.
- **OOD metrics**: p_oodc and p_re are the targets — SAM should help these most.
- **Training curve**: Should be smoother than baseline.

## Baseline

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| p_tan  | **28.341** | < 28.34 |
| p_re   | **6.300**  | < 6.30  |

W&B: `hgml7i2r` (s42), `qic03vrg` (s73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent frieren --wandb_name "frieren/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT frieren:

## Results

### Metrics (mae_surf_p)

| Metric | Baseline (PR #2213) | Seed 42 | Seed 73 | 2-seed Avg | Delta |
|--------|---------------------|---------|---------|------------|-------|
| **p_in** | **11.979** | 18.267 | 22.284 | 20.276 | **+69.3%** |
| **p_tan** | **28.341** | 37.069 | 41.660 | 39.364 | **+38.9%** |
| **p_oodc** | **7.643** | 16.442 | 22.852 | 19.647 | **+157.1%** |
| **p_re** | **6.300** | 15.101 | 21.160 | 18.131 | **+187.8%** |

### W&B Run IDs
- Seed 42: `k9zjoibj` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/k9zjoibj)
- Seed 73: `o127w2zx` — [link](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/o127w2zx)

### Command used
```bash
cd cfd_tandemfoil && python train.py \
  --agent frieren --wandb_name "frieren/sam-s42" \
  --wandb_group "round19/sam-optimizer" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 80 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --sam --sam_rho 0.05
```
(Seed 73 identical with `--seed 73 --wandb_name "frieren/sam-s73"`)

Peak memory: ~60GB (GPU 0), ~76GB (GPU 1)

### Epoch count and training time
- **Total epochs: 86** (both seeds) — roughly half the baseline's ~150 due to 2x compute per step
- Best epoch: 85 (both seeds) — model was still improving at timeout, indicating severe underconvergence
- `cosine_T_max=80` schedule completed, but the model needed more epochs

### What happened

**Very strong negative result.** All metrics regressed catastrophically, with OOD metrics worst (+157-188%).

**Why it failed:**

1. **2x compute cost is devastating in a fixed time budget.** SAM requires two forward+backward passes per step, halving the epoch count from ~150 to ~86. The model was still improving at epoch 85 (best_epoch = 85 = last epoch), meaning it didn't converge at all. The baseline needs ~120-150 epochs for good convergence.

2. **SAM skipped when PCGrad is active.** The existing SAM implementation skips the perturbation step when PCGrad is running (`not use_pcgrad` guard at line 2252). Since `--pcgrad_3way` is enabled and PCGrad is active for most batches (any batch with both in-dist and OOD samples), SAM only ran on the subset of batches where PCGrad wasn't triggered. This means SAM's benefit was inconsistently applied.

3. **cosine_T_max=80 too short for remaining epochs.** With the cosine schedule completing at epoch 80, the model spent the last 6 epochs at near-minimum LR (~1e-5), but it had only just started learning surface features (surface weight ramps up over 60 epochs, and with SAM's slower convergence, this ramp needed even longer).

4. **SAM + Lion may be suboptimal.** SAM's perturbation scales gradients by `rho / ||g||`, then Lion takes the sign of the perturbed gradient. The sign-based update may not preserve SAM's carefully calibrated perturbation direction — Lion discards gradient magnitude information that SAM relies on.

5. **High inter-seed variance (s42: p_in=18.3 vs s73: p_in=22.3).** The 4 Pa gap between seeds suggests the halved epoch budget puts training in a high-variance convergence regime.

### Implementation note

Fixed a bug in the existing SAM second-pass code: `loss2.backward()` was failing with "trying to backward through the graph a second time" because shared tensors (`x`, `y_norm`) from the first pass retained autograd history. Fixed by detaching `x` and `y_norm` before the second forward pass.

### Suggested follow-ups

- **SAM is not viable with a 3-hour training budget.** The 2x compute cost reduces epochs by ~50%, and this model needs all 150 epochs to converge. SAM would need a 6-hour budget to be fairly tested.
- **If SAM is revisited**: use `adaln_sam` mode (last 25% only) which already existed. This gives full convergence first, then SAM refinement. But note that the existing `adaln_sam` flag was already available and untested against the current baseline.
- **Alternatively: Efficient SAM (ESAM/SAM-ON)** — variants like mSAM (mini-batch SAM) or ESAM that approximate SAM with ~1.2-1.5x compute instead of 2x. This would preserve more epochs.
- **Focus on approaches that don't increase per-step compute.** Given the fixed time budget, any technique that increases per-step cost proportionally reduces convergence. The baseline cosine schedule + EMA is already well-optimized for this budget.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2242 because ADVISOR: Closing — SAM optimizer is catastrophic across all metrics.

Results (2-seed avg vs baseline):
- p_in: 20.276 vs 11.979 (+69.3%)
- p_oodc: 19.647 vs 7.643 (+157.1%)
- p_tan: 39.364 vs 28.341 (+38.9%)
- p_re: 18.131 vs 6.300 (+187.8%)

Root causes:
1. SAM doubles forward/backward passes → only 86 epochs in wall-clock budget (vs 145 baseline)
2. SAM skipped when PCGrad active → inconsistent optimization
3. cosine_T_max=80 too short for 86-epoch run
4. SAM + Lion sign-based updates may be fundamentally incompatible — SAM perturbs in steepest direction, but Lion discards gradient magnitude

Dead end — SAM is infeasible within our wall-clock and optimizer constraints..
